From 56910ad70300775739cf00d7e75f8e58c2ae9406 Mon Sep 17 00:00:00 2001
From: Andrzej Pietrasiewicz <andrzej.p@samsung.com>
Date: Tue, 14 Jan 2014 14:27:13 +0100
Subject: [PATCH 1100/1302] Revert "usb/gadget: f_mass_storage: use
 fsg_common_set_nluns in fsg_common_init"

This reverts commit 988ececeab73edfa2e4f27ad7167034d8908aa74.

Signed-off-by: MyungJoo Ham <myungjoo.ham@samsung.com>
---
 drivers/usb/gadget/f_mass_storage.c | 22 +++++++++++++++++-----
 1 file changed, 17 insertions(+), 5 deletions(-)

diff --git a/drivers/usb/gadget/f_mass_storage.c b/drivers/usb/gadget/f_mass_storage.c
index 691fb51..2a71540 100644
--- a/drivers/usb/gadget/f_mass_storage.c
+++ b/drivers/usb/gadget/f_mass_storage.c
@@ -3006,6 +3006,12 @@ struct fsg_common *fsg_common_init(struct fsg_common *common,
 	int nluns, i, rc;
 	char *pathbuf;
 
+	/* Find out how many LUNs there should be */
+	nluns = cfg->nluns;
+	if (nluns < 1 || nluns > FSG_MAX_LUNS) {
+		dev_err(&gadget->dev, "invalid number of LUNs: %u\n", nluns);
+		return ERR_PTR(-EINVAL);
+	}
 
 	common = fsg_common_setup(common, !!common);
 	if (IS_ERR(common))
@@ -3035,12 +3041,17 @@ struct fsg_common *fsg_common_init(struct fsg_common *common,
 	}
 	fsg_intf_desc.iInterface = us[FSG_STRING_INTERFACE].id;
 
-
-	rc = fsg_common_set_nluns(common, cfg->nluns);
-	if (rc)
+	/*
+	 * Create the LUNs, open their backing files, and register the
+	 * LUN devices in sysfs.
+	 */
+	curlun_it = kcalloc(nluns, sizeof(*curlun_it), GFP_KERNEL);
+	if (unlikely(!curlun_it)) {
+		rc = -ENOMEM;
 		goto error_release;
-	curlun_it = common->luns;
-	nluns = cfg->nluns;
+	}
+	common->luns = curlun_it;
+
 	for (i = 0, lcfg = cfg->luns; i < nluns; ++i, ++curlun_it, ++lcfg) {
 		struct fsg_lun *curlun;
 
@@ -3104,6 +3115,7 @@ struct fsg_common *fsg_common_init(struct fsg_common *common,
 			goto error_luns;
 		}
 	}
+	common->nluns = nluns;
 
 
 	/* Prepare inquiryString */
-- 
1.8.3.2

